health-services/ExerciseSampleCompose/app/src/main/java/com/example/exercisesamplecompose/presentation/ExerciseViewModel.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.example.exercisesamplecompose.presentation
+
+import android.Manifest
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.exercisesamplecompose.data.HealthServicesRepository
+import com.example.exercisesamplecompose.data.ServiceState
+import dagger.hilt.android.lifecycle.HiltViewModel
+import javax.inject.Inject
+import kotlinx.coroutines.launch
+
+
+@HiltViewModel
+class ExerciseViewModel @Inject constructor(
+    private val healthServicesRepository: HealthServicesRepository
+) : ViewModel() {
+
+    val permissions = arrayOf(
+        Manifest.permission.BODY_SENSORS,
+        Manifest.permission.ACCESS_FINE_LOCATION,
+        Manifest.permission.ACTIVITY_RECOGNITION
+    )
+
+    var hasExerciseCapabilities = mutableStateOf(true)
+    var isTrackingAnotherExercise = mutableStateOf(false)
+
+    private var _exerciseServiceState: MutableState<ServiceState> =
+        healthServicesRepository.exerciseServiceState
+    val exerciseServiceState = _exerciseServiceState
+
+    init {
```

Comment:
Launching coroutines in the `init` block of a VM can cause a subtle bug when it's used to update Compose state. This is because the coroutine may suspend on another thread before updating the Compose state, and the internals of it (the snapshot system) may not be ready for updates yet.

We can instead do this call lazily, instead of on initialization using the `flow { }` builder.


```kotlin
data class ExerciseUiState(
    val hasExerciseCapabilities: Boolean = true,
    val isTrackingAnotherExercise: Boolean = false,
)

class ExerciseViewModel {
    val uiState = flow {
        ExerciseUiState(
            hasExerciseCapabilities = healthServicesRepository.hasExerciseCapability(),
            isTrackingAnotherExercise = healthServicesRepository. isTrackingExerciseInAnotherApp(),
        )
    }.stateIn(
            viewModelScope,
            SharingStarted.whileSubscribed(3_000),
            ExerciseUiState()
        )
}
```

Reply:
Have implemented this. Will probably keep working on a way to get ExerciseCapabilities loading consistently, but for now it works